pkg/apis/pipeline/v1beta1/when_validation.go

```diff
@@ -0,0 +1,85 @@
+/*
+Copyright 2020 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"knative.dev/pkg/apis"
+)
+
+var validWhenOperators = []string{
+	string(selection.In),
+	string(selection.NotIn),
+}
+
+func (wes WhenExpressions) validate() *apis.FieldError {
+	if err := wes.validateWhenExpressionsFields(); err != nil {
+		return err
+	}
+	if err := wes.validateTaskResultsVariables(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (wes WhenExpressions) validateWhenExpressionsFields() *apis.FieldError {
+	for _, we := range wes {
+		if err := we.validateWhenExpressionFields(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (we *WhenExpression) validateWhenExpressionFields() *apis.FieldError {
+	if equality.Semantic.DeepEqual(we, &WhenExpression{}) || we == nil {
+		return apis.ErrMissingField(apis.CurrentField)
+	}
+	if we.Input == "" { //TODO(jerop): could someone use "" as a valid input?
```

Comment:
whenever we include TODOs we should reference an open issue to track the work - or deal with it before merging and remove the TODO

feels like it could be an okay input to me? i could imagine having a task produce a result that is an empty string (for some reason)

Reply:
`Input` is a string so its zero value is empty string -- `Input` defaults to empty string when not specified. Looked into disallowing defaulting to an empty string and I didn't find a good solution, please let me know if there's a good way to do this. The alternatives are:
1) documenting that if `Input` is not specified, it'd default to empty string and we'll compare that against the `Values`
2) disallowing empty strings as `Inputs` -- although some use cases could need this to be valid 

So maybe we can go with option 1 instead?

Reply:
good catch, very thorough analysis!! 🤩